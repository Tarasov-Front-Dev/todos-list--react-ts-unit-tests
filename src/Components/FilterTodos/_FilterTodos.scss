.todos__filter {
  @include main_border;
  padding: 7px 15px;
  display: flex;
  flex-flow: row nowrap;
  justify-content: space-between;
  align-items: center;
  font-size: smaller;
  background-color: white;

  .todos__filters {
    display: flex;
    flex-flow: row nowrap;
    gap: 15px;

    li {
      @include main_border;
      @extend %hover_active;

      padding: 8px;
      border-radius: 5px;
      border-color: transparent;
      cursor: pointer;

      &.toggled {
        border-color: $bgColor;
      }
    }
  }

  .todos__clearCompleted {
    @include main_border;
    @extend %hover_active;

    padding: 8px;
    border-radius: 5px;
    border-color: transparent;
    cursor: pointer;
  }
}
