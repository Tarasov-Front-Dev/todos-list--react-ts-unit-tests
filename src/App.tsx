import { useMemo, useState } from "react";
import FormNewTodo from "./Components/FormNewTodo/FormNewTodo";
import Todos from "./Components/TodosList/TodosList";
import FilterTodos, {
  FilterOptions,
} from "./Components/FilterTodos/FilterTodos";
import { filteredTodos } from "./Utils/Functions";

export type TodoModel = {
  id: number;
  value: string;
  completed: boolean;
};

function App() {
  const [todos, setTodos] = useState<TodoModel[]>([]);
  const [filter, setFilter] = useState(FilterOptions.ALL);

  const memoizedFilteredTodos: TodoModel[] = useMemo(
    () => filteredTodos(todos, filter),
    [todos, filter]
  );

  return (
    <article className="todos__wrapper">
      <h1>todos</h1>
      <FormNewTodo setTodos={setTodos} />
      <Todos todos={memoizedFilteredTodos} setTodos={setTodos} />
      <FilterTodos
        filter={filter}
        setFilter={setFilter}
        setTodos={setTodos}
        itemsLeft={memoizedFilteredTodos.length}
      />
    </article>
  );
}

export default App;
