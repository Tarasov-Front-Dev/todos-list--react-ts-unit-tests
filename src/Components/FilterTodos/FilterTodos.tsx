import { TodoModel } from "../../App";

export enum FilterOptions {
  ALL = "all",
  ACTIVE = "active",
  COMPLETED = "completed",
}

type FilterTodosProps = {
  filter: FilterOptions;
  setFilter: React.Dispatch<React.SetStateAction<FilterOptions>>;
  setTodos: React.Dispatch<React.SetStateAction<TodoModel[]>>;
  itemsLeft: number;
};

export default function FilterTodos({
  filter,
  setFilter,
  setTodos,
  itemsLeft,
}: FilterTodosProps) {
  const handleFilterChange = (value: string | null) => {
    if (!Object.keys(FilterOptions).some((key) => key === value?.toUpperCase()))
      return;
    else
      setFilter(
        FilterOptions[value?.toUpperCase() as keyof typeof FilterOptions]
      );
  };

  const handleClearCompletedTodos = () => {
    setTodos((todos) => todos.filter((todo) => !todo.completed));
    setFilter(FilterOptions.ALL);
  };

  const filterOptions = Object.values(FilterOptions).map((option) => (
    <li
      key={option}
      className={option === filter ? "toggled" : ""}
      onClick={(e) =>
        handleFilterChange((e.target as HTMLLIElement).textContent)
      }
    >
      {option[0].toUpperCase() + option.slice(1)}
    </li>
  ));

  return (
    <section className="todos__filter">
      <div className="todos__left">{itemsLeft} items left</div>
      <ul className="todos__filters">{filterOptions}</ul>
      <div
        className="todos__clearCompleted"
        onClick={handleClearCompletedTodos}
      >
        Clear completed
      </div>
    </section>
  );
}
