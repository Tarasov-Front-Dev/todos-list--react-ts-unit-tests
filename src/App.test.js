import { render, screen, fireEvent } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import App, { TodoModel } from "./App";
import { act } from "react-dom/test-utils";
import { FilterOptions } from "./Components/FilterTodos/FilterTodos";

const todosTruthy = [
  { id: 1, value: "First todo", completed: false },
  { id: 2, value: "Second todo", completed: true },
];

describe("App", () => {
  it("App renders", () => {
    render(<App />);

    expect(screen.getByText(/todos/i)).toBeInTheDocument();
    expect(
      screen.getByPlaceholderText(/What needs to be done?/i)
    ).toBeInTheDocument();
    expect(screen.getByText(/items left/)).toBeInTheDocument();
    expect(screen.getByText(/All/)).toBeInTheDocument();
    expect(screen.getByText(/Active/)).toBeInTheDocument();
    expect(screen.getByText(/Completed/)).toBeInTheDocument();
    expect(screen.getByText(/Clear completed/)).toBeInTheDocument();
  });

  it("Typing correctly", () => {
    render(<App />);
    expect(screen.queryByDisplayValue(/First task/)).toBeNull();

    act(() => userEvent.type(screen.getByRole("textbox"), "First task"));
    expect(screen.getByDisplayValue(/First task/)).toBeInTheDocument();
  });

  it("Adding todos correctly", () => {
    render(<App />);
    const form = screen.getByTestId("todos__form");
    const input = screen.getByRole("textbox");

    act(() => userEvent.type(input, "First task"));
    expect(input).toHaveDisplayValue("First task");
    act(() => fireEvent.submit(form));
    expect(input).not.toHaveDisplayValue("First task");

    act(() => userEvent.type(input, "Second task"));
    expect(input).toHaveDisplayValue("Second task");
    act(() => fireEvent.submit(form));
    expect(input).not.toHaveDisplayValue("Second task");
  });

  it("Completing todos correctly", () => {
    render(<App />);
    //--START--Initial state
    const form = screen.getByTestId("todos__form");
    const input = screen.getByRole("textbox");

    act(() => userEvent.type(input, "First task"));
    act(() => fireEvent.submit(form));

    act(() => userEvent.type(input, "Second task"));
    act(() => fireEvent.submit(form));

    const todosList = screen.getByTestId("todos__list");
    //---END--Initial state

    expect(todosList.children[0]).toContainHTML("Second task");
    expect(todosList.children[1]).toContainHTML("First task");

    expect(todosList.children[1]).not.toHaveClass("completed");
    // // Marking "First task" as completed
    act(() => userEvent.click(screen.getByText("First task")));
    expect(todosList.children[1]).toHaveClass("completed");
  });

  it("Toggle todos filter", () => {
    const AppComponent = render(<App />);

    //--START--Initial state
    const form = screen.getByTestId("todos__form");
    const input = screen.getByRole("textbox");

    act(() => userEvent.type(input, "First task"));
    act(() => fireEvent.submit(form));

    act(() => userEvent.type(input, "Second task"));
    act(() => fireEvent.submit(form));

    const todosList = screen.getByTestId("todos__list");
    const todosFilter = screen.getByTestId("todos__filter");

    // // Checking if "First task" toBeInTheDocument() on filter 'ALL'
    expect(todosFilter.children[0]).toHaveClass("toggled");
    expect(todosFilter.children[1]).not.toHaveClass("toggled");
    expect(todosFilter.children[2]).not.toHaveClass("toggled");
    expect(screen.getByText("First task")).toBeInTheDocument();
    // // Marking "First task" as completed
    act(() => userEvent.click(screen.getByText("First task"))); // click <li>'First task'</li>
    expect(screen.getByText("First task")).toBeInTheDocument();
    //---END--Initial state

    expect(screen.getByText("2 items left")).toBeInTheDocument();

    act(() => userEvent.click(todosFilter.children[1])); // click --Active--
    expect(screen.getByText("Second task")).toBeInTheDocument();
    expect(screen.queryByText("First task")).not.toBeInTheDocument();
    expect(screen.getByText("1 items left")).toBeInTheDocument();

    act(() => userEvent.click(todosFilter.children[2])); // click --Completed--
    expect(screen.getByText("First task")).toBeInTheDocument();
    expect(screen.queryByText("Second task")).not.toBeInTheDocument();
    expect(screen.getByText("1 items left")).toBeInTheDocument();

    act(() => userEvent.click(screen.getByText(/clear completed/i))); // auto toggle to --All--
    expect(screen.queryByText("First task")).not.toBeInTheDocument();
    expect(screen.getByText("Second task")).toBeInTheDocument();
    expect(screen.getByText("1 items left")).toBeInTheDocument();

    act(() => userEvent.click(todosFilter.children[1])); // click --Active--
    expect(screen.queryByText("First task")).not.toBeInTheDocument();
    expect(screen.getByText("Second task")).toBeInTheDocument();
    expect(screen.getByText("1 items left")).toBeInTheDocument();

    expect(AppComponent).toMatchSnapshot();
  });
});
