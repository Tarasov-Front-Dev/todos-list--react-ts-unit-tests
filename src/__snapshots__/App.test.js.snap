// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App Toggle todos filter 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <article
        class="todos__wrapper"
      >
        <h1>
          todos
        </h1>
        <form
          class="todos__form"
          data-testid="todos__form"
          name="todos__form"
        >
          <label
            for="newTodo"
          >
            <input
              autocomplete="off"
              id="newTodo"
              name="newTodo"
              placeholder="What needs to be done?"
              value=""
            />
          </label>
        </form>
        <ul
          class="todos__list"
          data-testid="todos__list"
        >
          <li
            class="todo__item"
          >
            Second task
          </li>
        </ul>
        <section
          class="todos__filter"
        >
          <div
            class="todos__left"
          >
            1
             items left
          </div>
          <ul
            class="todos__filters"
            data-testid="todos__filter"
          >
            <li
              class=""
            >
              All
            </li>
            <li
              class="toggled"
            >
              Active
            </li>
            <li
              class=""
            >
              Completed
            </li>
          </ul>
          <div
            class="todos__clearCompleted"
          >
            Clear completed
          </div>
        </section>
      </article>
    </div>
  </body>,
  "container": <div>
    <article
      class="todos__wrapper"
    >
      <h1>
        todos
      </h1>
      <form
        class="todos__form"
        data-testid="todos__form"
        name="todos__form"
      >
        <label
          for="newTodo"
        >
          <input
            autocomplete="off"
            id="newTodo"
            name="newTodo"
            placeholder="What needs to be done?"
            value=""
          />
        </label>
      </form>
      <ul
        class="todos__list"
        data-testid="todos__list"
      >
        <li
          class="todo__item"
        >
          Second task
        </li>
      </ul>
      <section
        class="todos__filter"
      >
        <div
          class="todos__left"
        >
          1
           items left
        </div>
        <ul
          class="todos__filters"
          data-testid="todos__filter"
        >
          <li
            class=""
          >
            All
          </li>
          <li
            class="toggled"
          >
            Active
          </li>
          <li
            class=""
          >
            Completed
          </li>
        </ul>
        <div
          class="todos__clearCompleted"
        >
          Clear completed
        </div>
      </section>
    </article>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
